spec/java/util/stream/Stream.main.lsl

```diff
@@ -0,0 +1,1049 @@
+    }
+
+
+    fun *.toArray (@target self: Stream): array<Object>
+    {
+        result = this.storage;
+    }
+
+
+    fun *.toArray (@target self: Stream, generator: IntFunction): array<Object>
+    {
+        val generatedArray: array<Object> = action CALL(generator,[this.length]) as array<Object>;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _copyToArray_loop(i, generatedArray)
+        );
+        result = generatedArray;
+    }
+
+
+    @Phantom proc _copyToArray_loop (i: int, generatedArray: array<Object>): void
+    {
+        generatedArray[i] = this.storage[i];
+    }
+
+
+    fun *.reduce (@target self: Stream, identity: Object, accumulator: BinaryOperator): Object
+    {
+        if (accumulator == null || identity == null)
+            _throwNPE();
+
+        result = identity;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _accumulate_loop(i, accumulator, result)
+        );
+    }
+
+
+    @Phantom proc _accumulate_loop (i: int, accumulator: BinaryOperator, result: Object): void
+    {
+        result = action CALL(accumulator, [result, this.storage[i]]);
+    }
+
+
+    fun *.reduce (@target self: Stream, accumulator: BinaryOperator): Optional
+    {
+        if (accumulator == null)
+            _throwNPE();
+
+        var value: Object = null;
+
+        if (this.length > 0)
+        {
+            value = this.storage[0];
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 1, this.length, +1,
+                _accumulate_optional_loop(i, accumulator, value)
+            );
+        }
+
+        result = action DEBUG_DO("Optional.of(value)");
+    }
+
+
+    @Phantom proc _accumulate_optional_loop (i: int, accumulator: BinaryOperator, value: Object): void
+    {
+        value = action CALL(accumulator, [value, this.storage[i]]);
+    }
+
+
+    fun *.reduce (@target self: Stream, identity:Object, accumulator: BiFunction, combiner: BinaryOperator): Object
+    {
+        if (accumulator == null || identity == null || combiner == null)
+            _throwNPE();
+
+        result = identity;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _accumulate_with_biFunction_loop(i, accumulator, result)
+        );
+        // since this implementation is always sequential, we do not need to use the combiner
+    }
+
+
+    @Phantom proc _accumulate_with_biFunction_loop (i: int, accumulator: BiFunction, result: Object): void
+    {
+        result = action CALL(accumulator, [result, this.storage[i]]);
+    }
+
+
+    fun *.collect (@target self: Stream, supplier: Supplier, accumulator: BiConsumer, combiner: BiConsumer): Object
+    {
+        if (supplier == null || accumulator == null || combiner == null)
+            _throwNPE();
+        // since this implementation is always sequential, we do not need to use the combiner
```

Comment:
This comment originates from UtBot sources. I think it should be noted explicitly.

Reply:
Fixed